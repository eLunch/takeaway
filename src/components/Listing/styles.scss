.utility-bar {
    display: flex;
    flex-direction: row;
    padding-bottom: 8px;

    & .input {
        -webkit-appearance: none;
        appearance: none;
        border-radius: 2px;
        background:#fff no-repeat 16px 8px;
        background-image:url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAAyBpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvIiB4bWxuczp4bXBNTT0iaHR0cDovL25zLmFkb2JlLmNvbS94YXAvMS4wL21tLyIgeG1sbnM6c3RSZWY9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9zVHlwZS9SZXNvdXJjZVJlZiMiIHhtcDpDcmVhdG9yVG9vbD0iQWRvYmUgUGhvdG9zaG9wIENTNSBXaW5kb3dzIiB4bXBNTTpJbnN0YW5jZUlEPSJ4bXAuaWlkOkU5NEY0RTlFMTA4NzExRTM5RTEzQkFBQzMyRjkyQzVBIiB4bXBNTTpEb2N1bWVudElEPSJ4bXAuZGlkOkU5NEY0RTlGMTA4NzExRTM5RTEzQkFBQzMyRjkyQzVBIj4gPHhtcE1NOkRlcml2ZWRGcm9tIHN0UmVmOmluc3RhbmNlSUQ9InhtcC5paWQ6RTk0RjRFOUMxMDg3MTFFMzlFMTNCQUFDMzJGOTJDNUEiIHN0UmVmOmRvY3VtZW50SUQ9InhtcC5kaWQ6RTk0RjRFOUQxMDg3MTFFMzlFMTNCQUFDMzJGOTJDNUEiLz4gPC9yZGY6RGVzY3JpcHRpb24+IDwvcmRmOlJERj4gPC94OnhtcG1ldGE+IDw/eHBhY2tldCBlbmQ9InIiPz4DjA/RAAABK0lEQVR42pTSQUdEURjG8dOY0TqmPkGmRcqYD9CmzZAWJRHVRIa0iFYtM6uofYaiEW2SRJtEi9YxIklp07ZkWswu0v/wnByve7vm5ee8M+85zz1jbt9Os+WiGkYdYxjCOx5wgFeXUHmtBSzpcCGa+5BJTCjEP+0nKWAT8xqe4ArPGEEVC1hHEbs2oBwdXkM7mj/JLZrad437sCGHOfUtcziutuYu2v8XUFF/4f6vMK/YgAH1HxkBYV60AR31gxkBYd6xAeF3VzMCwvzOBpypX8V4yuFRzX2d2gD/l5yjH4fYQEnzkj4fae5rJulF2sMXVrAsaTWttRFu4Osb+1jEDT71/ZveyhouTch2fINQL9hKefKjuYFfuznXWzXMTabyrvfyIV3M4vhXgAEAUMs7K0J9UJAAAAAASUVORK5CYII=);
        border: 1px solid #ebebeb;
        flex: 1 1 auto;
        font-size: 14px;
        height: 35px;
        line-height: 18px;
        padding: 10px 0;
        text-indent: 48px;
        width: 100%;
    }

    & .select-box {
        display: flex;
        align-items: center;
        flex-direction: row;
        flex: 0 0 auto;
        padding-left: 64px;

        & .label {
            font-size: 14px;
            font-weight: bold;
            margin-right: 16px;
        }

        & .select-wrapper {
            position: relative;
            
            & .select {
                width:100%;
                margin:0;
                background:none;
                border: 1px solid #ebebeb;
                border-radius: 2px;
                outline: none;
                -webkit-box-sizing: border-box;
                -moz-box-sizing: border-box;
                box-sizing: border-box;
                appearance: none;
                -webkit-appearance: none;
                font-size: 14px;
                padding: 7px 30px 7px 16px;
            }

            &:after {
                content: "";
                position: absolute;
                width: 9px;
                height: 8px;
                top: 50%;
                right: 1em;
                margin-top:-4px;
                z-index: 2;
                background: url("data:image/svg+xml;charset=utf8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 12'%3E%3Cpolygon fill='rgb(102,102,102)' points='8,12 0,0 16,0'/%3E%3C/svg%3E") 0 0 no-repeat;  
                pointer-events:none;
            }
        }
    }
}

.empty-result {
    font-size: 24;
}

@media all and (max-width: 476px) {
    .utility-bar .select-box {
        padding-left: 16px;
    }
}

@media all and (max-width: 720px) {
    .utility-bar .select-box {
        padding-left: 32px;
    }
}