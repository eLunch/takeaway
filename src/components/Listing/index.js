import React, { useState, useEffect } from 'react';
import { SORT_LIST } from '../../constants';
import data from '../../assets/data';
import { sortingByField, defaultSorting } from '../../helpers';

import './styles.scss';

import Items from '../Items';

const Listing = () => {
    // Initial state
    const [restaurantsList, setRestaurantsList] = useState(defaultSorting(data.restaurants));
    const [sortByFavorite, setSortByFavorite] = useState([]);
    const [searchValue, setSearchValue] = useState("");
    const [sortByField, setSortByValue] = useState("bestMatch");

    // Searching
    const handleChange = event => {
        setSearchValue(event.target.value)
    };
    useEffect(() => {
        let results = data.restaurants.filter(item => item.name.toLowerCase().includes(searchValue));

        setRestaurantsList(defaultSorting([...results]));
    }, [searchValue]);

    // Sorting by Select
    const handleChangeSelect = event => {
        setSortByValue(event.target.value)
    };
    useEffect(() => {
        let results = sortingByField(restaurantsList, sortByField, sortByFavorite);

        setRestaurantsList([...results]);
    }, [sortByField]);

    // Sorting by Favorite
    const handleClickFavorite = name => {
        !sortByFavorite.includes(name) ?
        setSortByFavorite(sortByFavorite.concat([name])) :
        setSortByFavorite(sortByFavorite.filter(item => item !== name));
    };
    useEffect(() => {
        restaurantsList.map((item) => {
            sortByFavorite.includes(item.name) ?
                (item.isFavorite = 1) :
                (item.isFavorite = 0)
        });

        let results = defaultSorting(restaurantsList, sortByFavorite);

        setRestaurantsList([...results]);
    }, [sortByFavorite]);

    // console.log(sortByFavorite, searchValue, sortByField);

    return (
        <div className="wrapper">
            <div className="utility-bar">
                <input
                    type="search"
                    placeholder="Search for restaurants"
                    value={searchValue}
                    onChange={handleChange}
                    className="input"
                />
                <div className="select-box">
                    <label className="label">Sort by</label>
                    <div className="select-wrapper">
                        <select onChange={handleChangeSelect} className="select" value={sortByField}>
                            {SORT_LIST.map((item, key) => {
                                return (
                                    <option
                                        value={item.value}
                                        label={item.label}
                                        key={key + item.value}
                                    />
                                );
                            })}
                        </select>
                    </div>
                </div>
            </div>
            {
                restaurantsList.length !== 0 ?
                    <Items 
                        items={restaurantsList}
                        sortByFavorite={sortByFavorite}
                        handleClickFavorite={handleClickFavorite}
                    /> :
                    <h2 className="empty-result">No restaurants found. Try again.</h2>
            }
        </div>
    )
}

export default Listing;