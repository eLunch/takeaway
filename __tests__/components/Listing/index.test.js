import * as React from 'react';
import { shallow } from 'enzyme';

import Listing from '../../../src/components/Listing';

describe('Filters', () => {
    it('should render as expected ', () => {
        const component = shallow(
            <Listing />
        );
    
        expect(component).toMatchSnapshot();
    });
});