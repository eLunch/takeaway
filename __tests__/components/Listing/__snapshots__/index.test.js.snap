// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters should render as expected  1`] = `
<div
  className="wrapper"
>
  <div
    className="utility-bar"
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="Search for restaurants"
      type="search"
      value=""
    />
    <div
      className="select-box"
    >
      <label
        className="label"
      >
        Sort by
      </label>
      <div
        className="select-wrapper"
      >
        <select
          className="select"
          onChange={[Function]}
          value="bestMatch"
        >
          <option
            key="0bestMatch"
            label="Best matchâ€‹"
            value="bestMatch"
          />
          <option
            key="1newest"
            label="Newest"
            value="newest"
          />
          <option
            key="2ratingAverage"
            label="Rating average"
            value="ratingAverage"
          />
          <option
            key="3distance"
            label="Distance"
            value="distance"
          />
          <option
            key="4popularity"
            label="Popularity"
            value="popularity"
          />
          <option
            key="5averageProductPrice"
            label="Average product price"
            value="averageProductPrice"
          />
          <option
            key="6deliveryCosts"
            label="Delivery costs"
            value="deliveryCosts"
          />
          <option
            key="7minCost"
            label="Minimum Costs"
            value="minCost"
          />
          <option
            key="8topRestaurants"
            label="Top restaurants"
            value="topRestaurants"
          />
        </select>
      </div>
    </div>
  </div>
  <Items
    handleClickFavorite={[Function]}
    items={
      Array [
        Object {
          "isFavorite": 0,
          "name": "Tanoshii Sushi",
          "sortingValues": Object {
            "averageProductPrice": 1536,
            "bestMatch": 0,
            "deliveryCosts": 200,
            "distance": 1190,
            "minCost": 1000,
            "newest": 96,
            "popularity": 17,
            "ratingAverage": 4.5,
            "topRestaurants": 21766,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "Sushi One",
          "sortingValues": Object {
            "averageProductPrice": 1285,
            "bestMatch": 3,
            "deliveryCosts": 0,
            "distance": 1618,
            "minCost": 1200,
            "newest": 238,
            "popularity": 23,
            "ratingAverage": 4,
            "topRestaurants": 38499,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "Roti Shop",
          "sortingValues": Object {
            "averageProductPrice": 915,
            "bestMatch": 4,
            "deliveryCosts": 0,
            "distance": 2308,
            "minCost": 2000,
            "newest": 247,
            "popularity": 81,
            "ratingAverage": 4.5,
            "topRestaurants": 187863,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "Aarti 2",
          "sortingValues": Object {
            "averageProductPrice": 922,
            "bestMatch": 5,
            "deliveryCosts": 250,
            "distance": 1605,
            "minCost": 500,
            "newest": 153,
            "popularity": 44,
            "ratingAverage": 4.5,
            "topRestaurants": 71542,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "Indian Kitchen",
          "sortingValues": Object {
            "averageProductPrice": 1189,
            "bestMatch": 11,
            "deliveryCosts": 150,
            "distance": 2308,
            "minCost": 1300,
            "newest": 272,
            "popularity": 5,
            "ratingAverage": 4.5,
            "topRestaurants": 12729,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "CIRO 1939",
          "sortingValues": Object {
            "averageProductPrice": 1762,
            "bestMatch": 12,
            "deliveryCosts": 99,
            "distance": 3957,
            "minCost": 1300,
            "newest": 231,
            "popularity": 79,
            "ratingAverage": 4.5,
            "topRestaurants": 314365,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "De Amsterdamsche Tram",
          "sortingValues": Object {
            "averageProductPrice": 892,
            "bestMatch": 304,
            "deliveryCosts": 0,
            "distance": 2792,
            "minCost": 0,
            "newest": 131,
            "popularity": 0,
            "ratingAverage": 0,
            "topRestaurants": 892,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "Lunchpakketdienst",
          "sortingValues": Object {
            "averageProductPrice": 4465,
            "bestMatch": 306,
            "deliveryCosts": 500,
            "distance": 14201,
            "minCost": 5000,
            "newest": 259,
            "popularity": 0,
            "ratingAverage": 3.5,
            "topRestaurants": 4465,
          },
          "status": "open",
        },
        Object {
          "isFavorite": 0,
          "name": "Royal Thai",
          "sortingValues": Object {
            "averageProductPrice": 1492,
            "bestMatch": 2,
            "deliveryCosts": 150,
            "distance": 2639,
            "minCost": 2500,
            "newest": 133,
            "popularity": 44,
            "ratingAverage": 4.5,
            "topRestaurants": 117608,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Pizza Heart",
          "sortingValues": Object {
            "averageProductPrice": 1103,
            "bestMatch": 6,
            "deliveryCosts": 150,
            "distance": 2453,
            "minCost": 1500,
            "newest": 118,
            "popularity": 9,
            "ratingAverage": 4,
            "topRestaurants": 23180,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Mama Mia",
          "sortingValues": Object {
            "averageProductPrice": 912,
            "bestMatch": 7,
            "deliveryCosts": 0,
            "distance": 1396,
            "minCost": 1000,
            "newest": 250,
            "popularity": 6,
            "ratingAverage": 4,
            "topRestaurants": 9288,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Feelfood",
          "sortingValues": Object {
            "averageProductPrice": 902,
            "bestMatch": 8,
            "deliveryCosts": 150,
            "distance": 2732,
            "minCost": 1500,
            "newest": 163,
            "popularity": 31,
            "ratingAverage": 4.5,
            "topRestaurants": 85594,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Fes Patisserie",
          "sortingValues": Object {
            "averageProductPrice": 1214,
            "bestMatch": 14,
            "deliveryCosts": 150,
            "distance": 2302,
            "minCost": 1250,
            "newest": 77,
            "popularity": 3,
            "ratingAverage": 4,
            "topRestaurants": 8120,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Yvonne's Vispaleis",
          "sortingValues": Object {
            "averageProductPrice": 2557,
            "bestMatch": 15,
            "deliveryCosts": 150,
            "distance": 2909,
            "minCost": 1750,
            "newest": 150,
            "popularity": 3,
            "ratingAverage": 5,
            "topRestaurants": 11284,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Lale Restaurant & Snackbar",
          "sortingValues": Object {
            "averageProductPrice": 838,
            "bestMatch": 305,
            "deliveryCosts": 0,
            "distance": 2880,
            "minCost": 0,
            "newest": 73,
            "popularity": 0,
            "ratingAverage": 0,
            "topRestaurants": 838,
          },
          "status": "order ahead",
        },
        Object {
          "isFavorite": 0,
          "name": "Tandoori Express",
          "sortingValues": Object {
            "averageProductPrice": 1146,
            "bestMatch": 1,
            "deliveryCosts": 150,
            "distance": 2308,
            "minCost": 1300,
            "newest": 266,
            "popularity": 123,
            "ratingAverage": 4.5,
            "topRestaurants": 285030,
          },
          "status": "closed",
        },
        Object {
          "isFavorite": 0,
          "name": "Daily Sushi",
          "sortingValues": Object {
            "averageProductPrice": 1327,
            "bestMatch": 9,
            "deliveryCosts": 200,
            "distance": 1911,
            "minCost": 1000,
            "newest": 221,
            "popularity": 6,
            "ratingAverage": 4,
            "topRestaurants": 12793,
          },
          "status": "closed",
        },
        Object {
          "isFavorite": 0,
          "name": "Pamukkale",
          "sortingValues": Object {
            "averageProductPrice": 968,
            "bestMatch": 10,
            "deliveryCosts": 0,
            "distance": 2353,
            "minCost": 2000,
            "newest": 201,
            "popularity": 25,
            "ratingAverage": 4,
            "topRestaurants": 59793,
          },
          "status": "closed",
        },
        Object {
          "isFavorite": 0,
          "name": "Zenzai Sushi",
          "sortingValues": Object {
            "averageProductPrice": 1579,
            "bestMatch": 13,
            "deliveryCosts": 0,
            "distance": 2911,
            "minCost": 2000,
            "newest": 155,
            "popularity": 36,
            "ratingAverage": 4,
            "topRestaurants": 106375,
          },
          "status": "closed",
        },
      ]
    }
    sortByFavorite={Array []}
  />
</div>
`;
